File: lab6/C++/QuickSort.cpp
#include <iostream>
#include <vector>

// Быстрая сортировка (Quick sort) с использованием последнего элемента в качестве опорного
static int partition(std::vector<int>& a, int low, int high) {
    int pivot = a[high];
    int i = low - 1;
    for (int j = low; j < high; ++j) {
        if (a[j] <= pivot) {
            ++i;
            std::swap(a[i], a[j]);
        }
    }
    std::swap(a[i + 1], a[high]);
    return i + 1;
}

static void quickSort(std::vector<int>& a, int low, int high) {
    if (low < high) {
        int pi = partition(a, low, high);
        quickSort(a, low, pi - 1);
        quickSort(a, pi + 1, high);
    }
}

static void printArray(const std::vector<int>& a) {
    for (int v : a) std::cout << v << ' ';
    std::cout << '\n';
}

int main() {
    std::vector<int> array{21, 4, 18, 6, 30, 2};
    std::cout << "Исходный массив:\n";
    printArray(array);

    quickSort(array, 0, static_cast<int>(array.size()) - 1);

    std::cout << "Отсортированный массив (Quick Sort):\n";
    printArray(array);
    return 0;
}

File: lab6/Python/QuickSort.py
"""Быстрая сортировка (Quick Sort).
Рекурсивная реализация с выбором опорного элемента — последний элемент в сегменте.
"""


def quick_sort(array, low, high):
    if low < high:
        # pi - это индекс разбиения, array[pi] находится на своем месте
        pi = partition(array, low, high)

        # Отсортировать элементы до и после разбиения
        quick_sort(array, low, pi - 1)
        quick_sort(array, pi + 1, high)

def partition(array, low, high):
    # Выбираем последний элемент в качестве опорного
    pivot = array[high]
    i = low - 1  # Индекс меньшего элемента

    for j in range(low, high):
        # Если текущий элемент меньше или равен опорному
        if array[j] <= pivot:
            i += 1

            # Поменять местами array[i] и array[j]
            array[i], array[j] = array[j], array[i]

    # Поменять местами array[i+1] и array[high] (опорный элемент)
    array[i + 1], array[high] = array[high], array[i + 1]

    return i + 1

def print_array(array):
    print(" ".join(map(str, array)))

# Основной метод для запуска сортировки
if __name__ == "__main__":
    array = [33, 4, 99, 17, 2, 48]

    print("Исходный массив:")
    print_array(array)

    quick_sort(array, 0, len(array) - 1)

    print("\nОтсортированный массив:")
    print_array(array)